File: android/src/main/java/com/sixmon/capacitordocumentscanner/DocumentScannerPlugin.java
package com.sixmon.capacitordocumentscanner;

import android.app.Activity;
import android.content.Intent;
import android.content.IntentSender;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.util.Base64;
import android.util.Log;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.vision.documentscanner.GmsDocumentScanner;
import com.google.mlkit.vision.documentscanner.GmsDocumentScannerOptions;
import com.google.mlkit.vision.documentscanner.GmsDocumentScanning;
import com.google.mlkit.vision.documentscanner.GmsDocumentScanningResult;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * This class is used to start a document scan using Google ML Kit.
 * It accepts parameters used to customize the document scan, and callback parameters.
 */
@CapacitorPlugin(name = "DocumentScanner", requestCodes = {9999})
public class DocumentScannerPlugin extends Plugin {

    private static final String TAG = "DocumentScannerPlugin";
    private static final int REQUEST_CODE_SCAN = 9999;
    
    // Response type constants
    private static final String RESPONSE_TYPE_IMAGE_FILE_PATH = "imageFilePath";
    private static final String RESPONSE_TYPE_BASE64 = "base64";
    
    // Scanner mode constants
    private static final String SCANNER_MODE_FULL = "FULL";
    private static final String SCANNER_MODE_BASE = "BASE";
    private static final String SCANNER_MODE_BASE_WITH_FILTER = "BASE_WITH_FILTER";
    
    // Store parameters for use in result handling
    private String currentResponseType = RESPONSE_TYPE_IMAGE_FILE_PATH;
    private int currentQuality = 100;

    /**
     * start the document scanner using Google ML Kit
     *
     * @param call contains JS inputs and lets you return results
     */
    @PluginMethod
    public void scanDocument(PluginCall call) {
        
        // Get configuration options from the call
        Integer maxNumDocuments = call.getInt("maxNumDocuments", 24);
        String scannerMode = call.getString("scannerMode", SCANNER_MODE_FULL);
        String responseType = call.getString("responseType", RESPONSE_TYPE_IMAGE_FILE_PATH);
        Integer quality = call.getInt("croppedImageQuality", 100);
        Boolean letUserAdjustCrop = call.getBoolean("letUserAdjustCrop", true);
        
        // Store parameters for use in result handling
        this.currentResponseType = responseType;
        this.currentQuality = quality != null ? quality : 100;
        
        // Build ML Kit scanner options
        GmsDocumentScannerOptions.Builder optionsBuilder = new GmsDocumentScannerOptions.Builder();
        
        // Set scanner mode
        if (SCANNER_MODE_BASE.equals(scannerMode)) {
            optionsBuilder.setScannerMode(GmsDocumentScannerOptions.SCANNER_MODE_BASE);
        } else if (SCANNER_MODE_BASE_WITH_FILTER.equals(scannerMode)) {
            optionsBuilder.setScannerMode(GmsDocumentScannerOptions.SCANNER_MODE_BASE_WITH_FILTER);
        } else {
            optionsBuilder.setScannerMode(GmsDocumentScannerOptions.SCANNER_MODE_FULL);
        }
        
        // Set result formats
        optionsBuilder.setResultFormats(
            GmsDocumentScannerOptions.RESULT_FORMAT_JPEG,
            GmsDocumentScannerOptions.RESULT_FORMAT_PDF
        );
        
        // Set page limit
        if (maxNumDocuments != null && maxNumDocuments > 0) {
            optionsBuilder.setPageLimit(maxNumDocuments);
        }
        
        // Set gallery import allowed (equivalent to letUserAdjustCrop)
        optionsBuilder.setGalleryImportAllowed(letUserAdjustCrop);
        
        GmsDocumentScannerOptions options = optionsBuilder.build();
        GmsDocumentScanner scanner = GmsDocumentScanning.getClient(options);
        
        // Get the scanner intent
        scanner.getStartScanIntent(getActivity())
            .addOnSuccessListener(intentSender -> {
                try {
                    // Save the call to process the result later
                    saveCall(call);

                    // Use traditional startIntentSenderForResult with declared request code
                    getActivity().startIntentSenderForResult(
                        intentSender,
                        REQUEST_CODE_SCAN, // Use the request code declared in annotation
                        null, // No fill-in intent
                        0,    // flagsMask
                        0,    // flagsValues
                        0     // extraFlags
                    );

                } catch (Exception e) {
                    call.reject("Failed to start document scanner: " + e.getMessage());
                }
            })
            .addOnFailureListener(e -> {
                call.reject("Failed to initialize document scanner: " + e.getMessage());
            });
    }

    @Override
    protected void handleOnActivityResult(int requestCode, int resultCode, Intent data) {
        super.handleOnActivityResult(requestCode, resultCode, data);
        
        if (requestCode != REQUEST_CODE_SCAN) {
            Log.d(TAG, "Request code mismatch - expected: " + REQUEST_CODE_SCAN + ", got: " + requestCode);
            return;
        }
        
        PluginCall call = getSavedCall();
        if (call == null) {
            Log.e(TAG, "No saved call found - this should not happen");
            return;
        }

        JSObject response = new JSObject();
        
        if (resultCode == Activity.RESULT_CANCELED) {
            // User cancelled the scan
            response.put("status", "cancel");
            call.resolve(response);
            return;
        }
        
        if (resultCode != Activity.RESULT_OK) {
            // Error occurred
            call.reject("Document scanning failed with result code: " + resultCode);
            return;
        }
        
        if (data == null) {
            call.reject("No data returned from document scanner");
            return;
        }
        
        // Extract scanning result
        GmsDocumentScanningResult scanningResult = GmsDocumentScanningResult.fromActivityResultIntent(data);
        if (scanningResult == null) {
            call.reject("Failed to extract scanning result");
            return;
        }
        
        // Process the scanned pages
        List<GmsDocumentScanningResult.Page> pages = scanningResult.getPages();
        if (pages == null || pages.isEmpty()) {
            call.reject("No pages were scanned");
            return;
        }

        try {
            ArrayList<String> scannedImages = new ArrayList<>();
            
            for (int i = 0; i < pages.size(); i++) {
                GmsDocumentScanningResult.Page page = pages.get(i);
                
                Uri imageUri = page.getImageUri();
                if (imageUri != null) {
                    String processedImage = processImage(imageUri, this.currentResponseType, this.currentQuality);
                    if (processedImage != null) {
                        scannedImages.add(processedImage);
                    }
                }
            }
            
            if (scannedImages.isEmpty()) {
                call.reject("Failed to process scanned images");
                return;
            }
            
            response.put("scannedImages", new JSArray(scannedImages));
            response.put("status", "success");
            
            call.resolve(response);
            
        } catch (Exception e) {
            call.reject("Failed to process scanned images: " + e.getMessage());
        }
    }
    
    /**
     * Process the scanned image based on response type
     *
     * @param imageUri URI of the scanned image
     * @param responseType "base64" or "imageUri"
     * @param quality JPEG quality (0-100)
     * @return processed image string
     */
    private String processImage(Uri imageUri, String responseType, int quality) {
        try {
            if (RESPONSE_TYPE_IMAGE_FILE_PATH.equals(responseType)) {
                // Use external files directory which is accessible and doesn't require permissions
                File picturesDir = getContext().getExternalFilesDir(android.os.Environment.DIRECTORY_PICTURES);
                if (picturesDir == null) {
                    // Fallback to external files directory root
                    picturesDir = getContext().getExternalFilesDir(null);
                }
                
                if (picturesDir == null) {
                    Log.e(TAG, "Failed to get external files directory");
                    return null;
                }
                
                // Create the directory if it doesn't exist
                if (!picturesDir.exists()) {
                    picturesDir.mkdirs();
                }
                
                // Copy to app's external files directory and return Capacitor-compatible URI
                File imageFile = new File(picturesDir, "scanned_" + System.currentTimeMillis() + ".jpg");
                copyUriToFile(imageUri, imageFile);
                
                // Make the file readable
                imageFile.setReadable(true, false);
                
                // Convert to Capacitor WebView URL that can be accessed from the web layer
                String fileUri = imageFile.toURI().toString();
                
                // Use Capacitor's bridge to get a web-accessible URL
                if (getBridge() != null && getBridge().getLocalUrl() != null) {
                    String webViewUrl = getBridge().getLocalUrl();
                    // Convert file path to a capacitor:// URL
                    String capacitorPath = fileUri.replace("file://", webViewUrl + "/_capacitor_file_");
                    return capacitorPath;
                }
                
                // Fallback to file:// URI if bridge is not available
                return fileUri;
            } else {
                // Convert to base64
                InputStream inputStream = getContext().getContentResolver().openInputStream(imageUri);
                if (inputStream == null) {
                    return null;
                }
                
                // Decode bitmap
                Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
                inputStream.close();
                
                if (bitmap == null) {
                    return null;
                }
                
                // Compress to JPEG
                ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
                bitmap.compress(Bitmap.CompressFormat.JPEG, quality, outputStream);
                byte[] imageBytes = outputStream.toByteArray();
                bitmap.recycle();
                
                // Encode to base64
                return Base64.encodeToString(imageBytes, Base64.NO_WRAP);
            }
        } catch (IOException e) {
            Log.e(TAG, "Error processing image", e);
            return null;
        }
    }
    
    /**
     * Copy content from URI to a file
     *
     * @param sourceUri source URI
     * @param destFile destination file
     */
    private void copyUriToFile(Uri sourceUri, File destFile) throws IOException {
        InputStream inputStream = getContext().getContentResolver().openInputStream(sourceUri);
        if (inputStream == null) {
            throw new IOException("Failed to open input stream");
        }
        
        FileOutputStream outputStream = new FileOutputStream(destFile);
        byte[] buffer = new byte[4096];
        int bytesRead;
        
        while ((bytesRead = inputStream.read(buffer)) != -1) {
            outputStream.write(buffer, 0, bytesRead);
        }
        
        outputStream.close();
        inputStream.close();
    }
}

File: ios/Plugin/DocumentScannerPlugin.swift
import Foundation
import Capacitor
import VisionKit

/**
 * This class contains functions that get called when
 * you use the DocumentScanner JavaScript functions
 */
@available(iOS 13.0, *)
@objc(DocumentScannerPlugin)
public class DocumentScannerPlugin: CAPPlugin, VNDocumentCameraViewControllerDelegate {

    // Response type constants
    private let RESPONSE_TYPE_IMAGE_FILE_PATH = "imageFilePath"
    private let RESPONSE_TYPE_BASE64 = "base64"

    // Store parameters for use in delegate callbacks
    private var savedCall: CAPPluginCall?
    private var currentResponseType: String = "imageFilePath"
    private var currentQuality: Int = 100

    /**
     * Opens the document scanner camera
     *
     * @param  call contains JS inputs and lets you return results
     */
    @objc func scanDocument(_ call: CAPPluginCall) {
        // Get configuration options from the call
        let responseType = call.getString("responseType") ?? RESPONSE_TYPE_IMAGE_FILE_PATH
        let quality = call.getInt("croppedImageQuality") ?? 100

        // Store parameters for use in delegate callbacks
        self.currentResponseType = responseType
        self.currentQuality = quality
        self.savedCall = call

        // Check if VisionKit is available
        guard #available(iOS 13.0, *) else {
            call.reject("Document scanning requires iOS 13.0 or later")
            return
        }

        // Create and present the document camera
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            let documentCameraViewController = VNDocumentCameraViewController()
            documentCameraViewController.delegate = self

            self.bridge?.viewController?.present(documentCameraViewController, animated: true, completion: nil)
        }
    }

    // MARK: - VNDocumentCameraViewControllerDelegate

    /**
     * Called when the user successfully scans one or more documents
     */
    public func documentCameraViewController(_ controller: VNDocumentCameraViewController, didFinishWith scan: VNDocumentCameraScan) {
        controller.dismiss(animated: true, completion: nil)

        guard let call = savedCall else {
            return
        }

        var scannedImages: [String] = []

        // Process each scanned page
        for pageIndex in 0..<scan.pageCount {
            let image = scan.imageOfPage(at: pageIndex)

            if currentResponseType == RESPONSE_TYPE_BASE64 {
                // Convert to base64
                if let base64String = convertImageToBase64(image, quality: currentQuality) {
                    scannedImages.append(base64String)
                }
            } else {
                // Save to file and return file path
                if let filePath = saveImageToDocumentsDirectory(image) {
                    scannedImages.append(filePath)
                }
            }
        }

        // Build response
        if scannedImages.isEmpty {
            call.reject("Failed to process scanned images")
        } else {
            call.resolve([
                "scannedImages": scannedImages,
                "status": "success"
            ])
        }

        savedCall = nil
    }

    /**
     * Called when the user cancels the document scan
     */
    public func documentCameraViewControllerDidCancel(_ controller: VNDocumentCameraViewController) {
        controller.dismiss(animated: true, completion: nil)

        guard let call = savedCall else {
            return
        }

        call.resolve([
            "status": "cancel"
        ])

        savedCall = nil
    }

    /**
     * Called when an error occurs during scanning
     */
    public func documentCameraViewController(_ controller: VNDocumentCameraViewController, didFailWithError error: Error) {
        controller.dismiss(animated: true, completion: nil)

        guard let call = savedCall else {
            return
        }

        call.reject("Document scanning failed: \(error.localizedDescription)")
        savedCall = nil
    }

    // MARK: - Helper Methods

    /**
     * Save a UIImage to the app's documents directory
     *
     * @param image The image to save
     * @return The file:// URL string, or nil if saving failed
     */
    private func saveImageToDocumentsDirectory(_ image: UIImage) -> String? {
        guard let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            return nil
        }

        let fileName = "scanned_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
        let fileURL = documentsDirectory.appendingPathComponent(fileName)

        // Convert image to JPEG data with quality
        let quality = CGFloat(currentQuality) / 100.0
        guard let imageData = image.jpegData(compressionQuality: quality) else {
            return nil
        }

        do {
            try imageData.write(to: fileURL)
            return fileURL.absoluteString
        } catch {
            print("Error saving image: \(error.localizedDescription)")
            return nil
        }
    }

    /**
     * Convert a UIImage to base64 string
     *
     * @param image The image to convert
     * @param quality JPEG quality from 0-100
     * @return The base64 encoded string, or nil if conversion failed
     */
    private func convertImageToBase64(_ image: UIImage, quality: Int) -> String? {
        let compressionQuality = CGFloat(quality) / 100.0
        guard let imageData = image.jpegData(compressionQuality: compressionQuality) else {
            return nil
        }

        return imageData.base64EncodedString()
    }
}